core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp

```diff
@@ -205,15 +215,47 @@ modify_launch_configuration_if_desired_occupancy_is_specified(
 
   if (dynamic_shmem > shmem) {
     shmem        = dynamic_shmem;
-    prefer_shmem = false;
+    prefer_shmem = CachePreference::PreferL1;
   }
 }
 
 template <class Policy>
 std::enable_if_t<!Policy::experimental_contains_desired_occupancy>
 modify_launch_configuration_if_desired_occupancy_is_specified(
-    Policy const&, cudaDeviceProp const&, cudaFuncAttributes const&,
-    dim3 const& /*block*/, int& /*shmem*/, bool& /*prefer_shmem*/) {}
+    dim3 const& grid, Policy const&, cudaDeviceProp const& properties,
+    cudaFuncAttributes const& attributes, dim3 const& block, int& shmem,
+    CachePreference& prefer_shmem) {
+  // Calculate maximum number of blocks that can simultaneously run on a SM
+  // based on the block size requested.
+  int const block_size   = block.x * block.y * block.z;
+  int max_blocks_threads = properties.maxThreadsPerMultiProcessor / block_size;
+  int const league_size  = grid.x * grid.y * grid.z;
+
+  // Calculate the maximum number of blocks that can simultaneously run based on
+  // the number of registers.
+  int const regs_per_sm     = properties.regsPerMultiprocessor;
+  int const regs_per_thread = attributes.numRegs;
+  int const max_blocks_regs = regs_per_sm / (regs_per_thread * block_size);
```

Comment:
This needs the fix to round up to the nearest multiple of 8. It might be good to package some bits of the occupancy calculation into utility functions that can get reused.

Reply:
I re-used the existing routine that does this calculation already.